### PR TITLE
Adds the Technocracy Advanced Cloning System Kit

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Adds the Technocracy Advanced Cloning System Kit to greys uplinks for 25 TC
A delicate cloner that consumes a lot of power to rapidly clone the person who was implanted with the biochip linked to it on death. 
Can be reused if chip is recovered, but easier said than done. Not in surplus, can be given to other people.

## Why It's Good For The Game

An interesting species item. Will work well for you if it is not found, but is going to be respawning you with almost nothing, a pain to reuse, and very vunerable. Probably works best in a team or with martial arts, but should work for you at least once.
After all, planning to fail is failure to plan.

## Testing

Killing and cloning myself a bunch.

## Changelog
:cl:
add: Adds the Technocracy Advanced Cloning System Kit to grey uplinks.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/grey_autocloner.dm [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	if(clonemind.active)	//somebody is using that mind
+		if(ckey(clonemind.key) != R.ckey)
+			return 0
+		if(clonemind.suicided) // Nah, if you are being dragged to the borg factory, that's on you
+			return 0
+	else
+		// get_ghost() will fail if they're unable to reenter their body
+		var/mob/dead/observer/G = clonemind.get_ghost()
+		if(!G)
+			return 0
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm:52]:
```suggestion
	if(attempting || stat & (NOPOWER|BROKEN))
		return FALSE
	clonemind = locateUID(R.mind)
	if(!istype(clonemind))	//not a mind
		return FALSE
	if(clonemind.current && clonemind.current.stat != DEAD)	//mind is associated with a non-dead body
		return FALSE
	if(clonemind.active)	//somebody is using that mind
		if(ckey(clonemind.key) != R.ckey)
			return FALSE
		if(clonemind.suicided) // Nah, if you are being dragged to the borg factory, that's on you
			return FALSE
	else
		// get_ghost() will fail if they're unable to reenter their body
		var/mob/dead/observer/G = clonemind.get_ghost()
		if(!G)
			return FALSE
```

File: code/game/objects/structures/grey_autocloner.dm [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+	H.Paralyse(8 SECONDS)
+
+	clonemind.transfer_to(H) //INSTANTLY INTO THE CLONE
+	H.ckey = R.ckey
+	update_clone_antag(H) //Since the body's got the mind, update their antag stuff right now. Otherwise, wait until they get kicked out (as per the CLONER_MATURE_CLONE business) to do it.
+	var/message
+	message += "<b>Consciousness slowly creeps over you as your body regenerates.</b><br>"
+	message += "<i>So this is what cloning feels like?</i>"
+	to_chat(H, "<span class='notice'>[message]</span>")
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm:89]:
```suggestion
	to_chat(H, "<span class='notice'><b>Consciousness slowly creeps over you as your body regenerates.</b><br><i>So this is what cloning feels like?</i></span>")
```

File: code/game/objects/structures/grey_autocloner.dm [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+	update_icon()
+
+	H.suiciding = FALSE
+	attempting = FALSE
+	addtimer(CALLBACK(src, PROC_REF(finish_clone)), 1 MINUTES)
+	return 1
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm:96]: consider not making them suffer the pains of being stuck in the machine for a minute
```suggestion
	return TRUE
```
Also is this isn't even being compared so you may want to just remove the returns

File: code/game/objects/structures/grey_autocloner.dm [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+		if(SSticker.mode.cult_risen)
+			SSticker.mode.rise(H)
+			if(SSticker.mode.cult_ascendant)
+				SSticker.mode.ascend(H)
+
+/obj/machinery/grey_autocloner/proc/finish_clone()
+	playsound(src.loc, 'sound/machines/ding.ogg', 50, 1)
+	var/obj/item/organ/internal/storedorgan = new /obj/item/organ/internal/cyberimp/brain/speech_translator
+	storedorgan.insert(occupant) //insert stored organ into the user
+	message_admins("fries are done")
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm]:
```suggestion
```

File: code/game/objects/structures/grey_autocloner.dm [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	if(H.mind in SSticker.mode.cult)
+		SSticker.mode.update_cult_icons_added(H.mind) // Adds the cult antag hud
+		SSticker.mode.add_cult_actions(H.mind) // And all the actions
+		if(SSticker.mode.cult_risen)
+			SSticker.mode.rise(H)
+			if(SSticker.mode.cult_ascendant)
+				SSticker.mode.ascend(H)
+
+/obj/machinery/grey_autocloner/proc/finish_clone()
+	playsound(src.loc, 'sound/machines/ding.ogg', 50, 1)
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm:124]:
```suggestion
	playsound(get_turf(src), 'sound/machines/ding.ogg', 50, TRUE)
```

File: code/game/objects/structures/grey_autocloner.dm [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+	message_admins("fries are done")
+	occupant.forceMove(get_turf(src))
+	occupant.update_body()
+	domutcheck(occupant) //Waiting until they're out before possible notransform.
+	occupant.special_post_clone_handling()
+	occupant = null
+	update_icon()
+
+/obj/machinery/grey_autocloner/emp_act(severity)
+	messy_explode()
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm:135]: uhhh you sure you want this to just explode if it gets graced by any emp?

File: code/game/objects/structures/grey_autocloner.dm [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+/obj/machinery/grey_autocloner/proc/messy_explode()
+	if(occupant)
+		occupant.forceMove(get_turf(src))
+	explosion(loc, 1, 2, 4, flame_range = 2)
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm:140]:
```suggestion
	explosion(get_turf(src), 1, 2, 4, flame_range = 2)
```

File: code/game/objects/structures/grey_autocloner.dm [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+	qdel(src)
+
+/obj/machinery/grey_autocloner/deconstruct(disassembled)
+	messy_explode()
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm:144]: not sure this will get called

**Review by GDNgit — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/grey_autocloner.dm (hunk 6 above)
Comment (Qwertytoforty) [reply code/game/objects/structures/grey_autocloner.dm:135]: Yes

File: code/datums/uplink_items/uplink_traitor.dm [hunk 9]
```diff
@@ -364,6 +364,16 @@
...
 	species = list("Human")
 
+//Gr(e)(a)y
+/datum/uplink_item/species_restricted/prescan
+	name = "Technocracy Advanced Cloning System"
+	desc = "This kit will give you the parts to build an advanced automatic cloning system, to clone whoever has the linked implant installed on death. \
+	Power intensive, implant must be recovered for reuse, and implanter must be linked to cloner."
+	reference = "TACS"
+	item = /obj/item/storage/box/syndie_kit/prescan
+	cost = 25 /// A fresh start, but a start with nothing. Hard to use well
```
Comment (lewcc) [inline code/datums/uplink_items/uplink_traitor.dm:374]:
```suggestion
	cost = 25 // A fresh start, but a start with nothing. Hard to use well
```

File: code/game/objects/structures/grey_autocloner.dm (hunk 3 above)
Comment (Qwertytoforty) [reply code/game/objects/structures/grey_autocloner.dm:96]: Nope. No meta knowledge after death, and the fun apprehention if someone is coming while you are being revived with no way to be seen

File: code/game/objects/structures/grey_autocloner.dm [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+			return
+
+	else if((occupant) && (occupant.loc == src))
+		occupant.Paralyse(8 SECONDS)
+		use_power(15000) //Hope the room has power!
+		for(var/bt in brine_types)
+			if(occupant.reagents.get_reagent_amount(bt) < 1)
+				occupant.reagents.add_reagent(bt, 1)
+
+/obj/machinery/grey_autocloner/proc/update_clone_antag(mob/living/carbon/human/H)
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/game/objects/structures/grey_autocloner.dm:111]: This is how the cloner does it

File: code/game/objects/items/weapons/implants/implant_prescan.dm [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+	if(linked)
+		linked.growclone(our_record)
+
+/obj/item/implant/grey_autocloner/implant(mob/source, mob/user, force)
+	if(!linked)
+		atom_say("Please link the implanter with a Technocracy cloning pod!")
+		return 0
+	. = ..()
+	if(.)
+		if(ishuman(imp_in))
```
Comment (lewcc) [inline code/game/objects/items/weapons/implants/implant_prescan.dm:22]: You could probably guard clause this on `if(!. || !ishuman(imp_in) return`

File: code/game/objects/structures/grey_autocloner.dm [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+	anchored = TRUE
+	density = TRUE
+	var/mob/living/carbon/human/occupant
+	var/datum/gas_mixture/air_contents = new()
+	var/list/brine_types = list("corazone", "perfluorodecalin", "epinephrine", "salglu_solution") //Taken from cloner, a bit of extra healing though they should be fully good.
+	var/datum/mind/clonemind
+	var/attempting = FALSE
+
+/obj/machinery/grey_autocloner/attackby(obj/item/implanter/implant, mob/user, params)
+	if(!istype(implant) || !if(istype(implant.imp, /obj/item/implant/grey_autocloner)))
```
Comment (GDNgit) [inline code/game/objects/structures/grey_autocloner.dm:15]:
```suggestion
	if(!istype(implant) || !(istype(implant.imp, /obj/item/implant/grey_autocloner)))
```

File: code/game/objects/structures/grey_autocloner.dm [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+	addtimer(CALLBACK(src, PROC_REF(finish_clone)), 1 MINUTES)
+	return 1
+
+/obj/machinery/grey_autocloner/process()
+	if(stat & NOPOWER) //explode if power is lost and cloning
+		if(occupant)
+			messy_explode()
+			return
+
+	else if((occupant) && (occupant.loc == src))
```
Comment (lewcc) [inline code/game/objects/structures/grey_autocloner.dm:99]:
```suggestion
	else if(occupant?.loc == src)
```

**Comment by Ionward** [comment]
technocracy mentioned 🥳

File: code/datums/uplink_items/uplink_traitor.dm (hunk 9 above)
Comment (1080pCat) [inline code/datums/uplink_items/uplink_traitor.dm:374]:
```suggestion
	cost = 25 /// A fresh start, but a start with nothing. Hard to use as well
```

**Comment by datlo** [comment]
Emag effect suggestion : If there's a user, make the cloner spits out a dead, brainless clone of the registered user. Can only be done once to prevent spam. Does not interfere with regular cloner usage, just consumes some power to make the clone.

This could be a fun bonus to fool security into thinking you got murdered by another antag.

**Comment by S34NW** [comment]
Please merge `upstream/master` into your branch to fix CI failing due to external factors.

File: code/game/objects/structures/grey_autocloner.dm [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+	. = ..()
+
+	air_contents.oxygen = MOLES_O2STANDARD * 2
+	air_contents.nitrogen = MOLES_N2STANDARD
+	air_contents.temperature = T20C
+
+/obj/machinery/grey_autocloner/Destroy() //These will only be not null if deleted with someone in it
+	occupant = null
+	clonemind = null
+	return ..()
```
Comment (Sirryan2002) [inline code/game/objects/structures/grey_autocloner.dm:31]: move these to right after the type definition

File: code/game/objects/structures/grey_autocloner.dm [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+	for(var/datum/language/L in R.languages)
+		H.add_language(L.name)
+
+	domutcheck(H, MUTCHK_FORCED) //Ensures species that get powers by the species proc handle_dna keep them
+
+	H.dna.UpdateSE()
+	H.dna.UpdateUI()
+
+	H.sync_organ_dna(1) // It's literally a fresh body as you can get, so all organs properly belong to it
```
Comment (Sirryan2002) [inline code/game/objects/structures/grey_autocloner.dm:76]:
```suggestion
	H.sync_organ_dna(TRUE) // It's literally a fresh body as you can get, so all organs properly belong to it
```

File: code/game/objects/structures/grey_autocloner.dm (hunk 10 above)
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/objects/structures/grey_autocloner.dm:106]: perhaps `transfer_to()` on the mind datum would be better

File: code/game/objects/structures/grey_autocloner.dm [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+	qdel(src)
+
+/obj/machinery/grey_autocloner/deconstruct(disassembled)
+	messy_explode()
+
+/obj/machinery/grey_autocloner/update_icon_state()
+	if(occupant)
+		icon_state = "experiment"
+	else
+		icon_state = "experiment-open"
```
Comment (Sirryan2002) [inline code/game/objects/structures/grey_autocloner.dm:145]:
```suggestion
	icon_state = occupant ? "experiment" : "experiment-open"
```

File: code/game/objects/items/weapons/implants/implant_prescan.dm [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+			our_record.name = B.dna.real_name
+			our_record.types = DNA2_BUF_UI|DNA2_BUF_UE|DNA2_BUF_SE
+			our_record.languages = imp_in.languages
+			if(imp_in.mind) //Save that mind so traitors can continue traitoring after cloning.
+				our_record.mind = "imp_in.mind.UID()"
+
+/obj/item/implant/grey_autocloner/Destroy()
+	linked = null
+	our_record = null
+	return ..()
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/game/objects/items/weapons/implants/implant_prescan.dm:39]: done

File: code/game/objects/structures/grey_autocloner.dm [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+	desc = "An advanced Technocracy cloning pod, used for rapid cloning. Very delicate, very power hungry, and very much should not be tampered with."
+	icon = 'icons/obj/abductor.dmi'
+	icon_state = "experiment-open"
+	anchored = TRUE
+	density = TRUE
+	var/mob/living/carbon/human/occupant
+	var/datum/gas_mixture/air_contents = new()
+	var/list/brine_types = list("corazone", "perfluorodecalin", "epinephrine", "salglu_solution") //Taken from cloner, a bit of extra healing though they should be fully good.
+	var/datum/mind/clonemind
+	var/attempting = FALSE
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/game/objects/structures/grey_autocloner.dm:12]: Done

File: code/game/objects/structures/grey_autocloner.dm [hunk 19]
```diff
@@ -0,0 +1,145 @@
+/obj/machinery/grey_autocloner
+	name = "technocracy cloning pod"
+	desc = "An advanced Technocracy cloning pod, used for rapid cloning. Very delicate, very power hungry, and very much should not be tampered with."
+	icon = 'icons/obj/abductor.dmi'
+	icon_state = "experiment-open"
+	anchored = TRUE
+	density = TRUE
+	var/mob/living/carbon/human/occupant
+	var/datum/gas_mixture/air_contents = new()
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/game/objects/structures/grey_autocloner.dm:9]: Done

File: code/game/objects/structures/grey_autocloner.dm (hunk 14 above)
Comment (Qwertytoforty) [reply code/game/objects/structures/grey_autocloner.dm:31]: done

File: code/game/objects/items/weapons/implants/implant_prescan.dm (hunk 11 above)
Comment (Qwertytoforty) [reply code/game/objects/items/weapons/implants/implant_prescan.dm:23]: Not allowing implantation with this

Comment (Qwertytoforty) [reply code/game/objects/items/weapons/implants/implant_prescan.dm:23]: I might be stupid, ignore me

Comment (Qwertytoforty) [reply code/game/objects/items/weapons/implants/implant_prescan.dm]: done

File: code/game/objects/structures/grey_autocloner.dm (hunk 10 above)
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/game/objects/structures/grey_autocloner.dm]: Done, removed from grey and base cloner

File: code/game/objects/structures/grey_autocloner.dm [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+/obj/machinery/grey_autocloner/process()
+	if(stat & NOPOWER) //explode if power is lost and cloning
+		if(occupant)
+			messy_explode()
+			return
+
+	else if(occupant?.loc == src)
+		occupant.Paralyse(8 SECONDS)
+		use_power(15000) //Hope the room has power!
```
Comment (DGamerL) [inline code/game/objects/structures/grey_autocloner.dm:101]: Might make it check if the room has enough power?
```suggestion
		if(!use_power(15000)) //Hope the room has power!
			return
```
Something like this

**Review by DGamerL — COMMENTED** [review]
Looks fine imo, but cloning is not my strong suit

**Review by Sirryan2002 — APPROVED** [review]
see Dgamer's comment

File: code/datums/uplink_items/uplink_traitor.dm [hunk 21]
```diff
@@ -365,6 +365,16 @@
 	cost = 10
 	species = list("Human")
 
+//Gr(e)(a)y
```
[the remark this answers is not in the document]
Comment (DGamerL) [reply code/datums/uplink_items/uplink_traitor.dm:368]: I was too silly, keep it 👍

**Review by S34NW — APPROVED** [review]

**Pull request merged**